backend/x11/output.c

```diff
@@ -229,11 +231,17 @@ static bool output_commit_buffer(struct wlr_x11_output *output) {
 	}
 
 	xcb_xfixes_region_t region = XCB_NONE;
+	pixman_region32_t damage;
+	pixman_region32_init(&damage);
+	pixman_region32_copy(&damage, &output->exposed);
+	pixman_region32_clear(&output->exposed);
 	if (output->wlr_output.pending.committed & WLR_OUTPUT_STATE_DAMAGE) {
-		pixman_region32_t *damage = &output->wlr_output.pending.damage;
+		pixman_region32_union(&damage, &damage, &output->wlr_output.pending.damage);
+	}
 
+	if (pixman_region32_not_empty(&damage)) {
```

Comment:
The compositor may not support damage tracking. In that case, the `WLR_OUTPUT_STATE_DAMAGE` flag won't be set and we want to damage the whole window (`region = XCB_NONE`).

If we got an Expose event (so the `expose` region isn't empty) but the compositor doesn't support damage tracking, we still want to damage the whole window. So we should probably merge this block into the previous `if`.

Reply:
Ah OK. I wasn't quite aware of the various permutations. So then we should only pay attention to the damage from events if "committed & WLR_OUTPUT_STATE_DAMAGE" is set, and otherwise not worry about it, since it'll be a full redraw anyways?

Reply:
Yeah, that sounds correct.